src/snowflake/cli/_plugins/nativeapp/entities/application_package.py

```diff
@@ -546,8 +565,73 @@ def _bundle(self):
         compiler.compile_artifacts()
+
+        # TODO should this merged into NativeAppCompiler?
+        self._inject_spcs(spcs_services)
+
         return bundle_map
 
+    def _inject_spcs(self, spcs_services: list[ServiceEntity]):
+        manifest_path = find_manifest_file(self.deploy_root)
+        manifest = yaml.safe_load(manifest_path.read_text())
+        if "configuration" not in manifest:
+            manifest["configuration"] = {}
+        existing_grant_callback = manifest["configuration"].get("grant_callback")
+        wrapper_grant_callback = "_spcs_generation.grant_callback"
+        manifest["configuration"]["grant_callback"] = wrapper_grant_callback
+        # TODO set default_web_endpoint in manifest?
+        if manifest_path.is_symlink():
+            manifest_path.unlink()
+        manifest_path.write_text(yaml.safe_dump(manifest, sort_keys=False))
+
+        generated_setup_script = self._spcs_grant_callback(
+            name=wrapper_grant_callback,
+            service=spcs_services[0]._entity_model,  # noqa SLF001
+            existing_grant_callback=existing_grant_callback,
+        )
+
+        setup_script_path = find_setup_script_file(self.deploy_root)
+        setup_script = setup_script_path.read_text()
+        if setup_script_path.is_symlink():
+            setup_script_path.unlink()
+        setup_script_path.write_text(setup_script + generated_setup_script)
+
+    @staticmethod
+    def _spcs_grant_callback(
+        name: str, service: ServiceEntityModel, existing_grant_callback: str
+    ):
+        return dedent(
+            f"""\
+            -- Begin generated SPCS services, this section is managed by the Snowflake CLI
+            create schema if not exists _spcs_generation;
+
+            create or replace procedure {name}(privileges array)
+            returns string
+            as $$
+            begin
+              {f'call {existing_grant_callback}(privileges);' if existing_grant_callback else ''}
+              if (array_contains('CREATE COMPUTE POOL'::variant, privileges)) then
+                 create compute pool if not exists {service.compute_pool}
+                     min_nodes = {service.min_nodes}
+                     max_nodes = {service.max_nodes}
+                     instance_family = {service.instance_family};
+              end if;
+              if (array_contains('BIND SERVICE ENDPOINT'::variant, privileges)) then
+                 create service if not exists _spcs_generation.{service.fqn.name}
+                  in compute pool {service.compute_pool}
+                  from specification_file = '{service.specification_file}';
+              end if;
+              return 'done';
+            end;
+            $$;
+
+            create application role if not exists _spcs_generation_role;
+            grant usage on procedure {name}(array) to application role _spcs_generation_role;
+
+            -- End generated SPCS services
+            """
```

Comment:
Do you see a way to reuse the same logic from `SpcsEntity.deploy()` or `ServiceEntity.deploy()` once it will be implemented?